### PR TITLE
Fix NASNet

### PR DESCRIPTION
**Pull request by taehoonlee**
This PR fixes NASNet discussed in #10172 and #9586.

@fchollet, @titu1994, Please note that:
- The `reduction_1_` in `_reduction_a_cell` was wrong, which results in incorrect weight files. @titu1994 must provide new weight files by using this PR and @fchollet must upload them again.
- The `skip_reduction` must be `True` for `NASNetLarge`.
- The layers `stem_conv1`, `stem_1`, and `stem_2` must always be included for ImageNet models.

### CODE REVIEW

**Comment by titu1994** [comment]
Got it. I'll update the code on my end and get weights by tomorrow. 

However, when you say "stem_conv1, stem_1 and stem_2" must always be present, I think they are useful only for training a new model. 

Since Keras applications are inference built, they don't require the auxiliary outputs, and therefore don't need to be included in the model?

**Comment by taehoonlee** [comment]
@titu1994, The layers `stem_conv1`, `stem_1`, and `stem_2` are not auxiliary layers. Please see [the original stem layers](https://github.com/tensorflow/models/blob/master/research/slim/nets/nasnet/nasnet.py#L248).

**Comment by titu1994** [comment]
@taehoonlee, got it. I confused the stem at the beginning of the network with the auxiliary branch. 

I'll get started on this asap. Thank you again for finding this fix.

**Comment by fchollet** [comment]
Thanks a lot for finding the issue and providing a fix. Should we disable imagenet weight loading as well (for the time being)?

**Comment by titu1994** [comment]
@taehoonlee A few points to note. The issue is not with the flag `skip_reduction` itself. It should definitely be set to False. The bug is on lines `p = p0 if not skip_reduction else p`. 

The correct statement is `p = p0 if skip_reduction else p`. With this fix, and setting `skip_reduction=False`, NASNet Large and NASNet mobile work as intended. My mistake was in mixing up what p0 and p meant for CIFAR and ImageNet variants of NASNet.

While I am currently building the weights again, I believe the current NASNet large weights are correct with this fix. Can you evaluate NASNet large with the current weights on the ImageNet validation set to confirm this? (I dont have the validation set, else I would run it myself).

Also, another update, tt will take 3 more hours to finish building all of the weights. If evaluations locally on some sample images work out properly, I will upload them to my repo after that, which will probably take another hour.

**Comment by taehoonlee** [comment]
@titu1994, I already confirmed that this PR achieved 3.8% top-5 error on the first 1000 samples of ImageNet validation set. There is no bug on `skip_reduction`. Please see [the original config](https://github.com/tensorflow/models/blob/master/research/slim/nets/nasnet/nasnet.py#L79).

**Comment by taehoonlee** [comment]
If you don't have the ImageNet validation set, the following codes might help you:
```python
import numpy as np
import tensorflow as tf
import tensornets as nets
import tensorflow_hub as hub

from keras.applications.imagenet_utils import decode_predictions
from keras.applications.nasnet import NASNetLarge, NASNetMobile, preprocess_input

url = 'https://tfhub.dev/google/imagenet'
model_name = 'nasnet_large'

img = nets.utils.load_img('cat.png', target_size=378, crop_size=331)

inputs = tf.placeholder(tf.float32, [None, 331, 331, 3])
tfhub = hub.Module("%s/%s/classification/1" % (url, model_name))
features = tfhub(inputs, signature="image_classification",
                 as_dict=True)
model_tfhub = tf.nn.softmax(features['default'])

model = NASNetLarge(input_shape=(331, 331, 3))
model.load_weights("/home/taehoonlee/.keras/models/NASNet-large.h5")

preds = model.predict(preprocess_input(img.copy()))
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    preds_tfhub = sess.run(model_tfhub, {inputs: img / 255.})
np.testing.assert_allclose(preds, preds_tfhub[:, 1:], atol=1e-4)
```

Note that the file `NASNet-large.h5` is not from `github.com/fchollet/deep-learning-models/releases` but from `github.com/titu1994/Keras-NASNet/releases`.

**Comment by titu1994** [comment]
> reduction_1_ in _reduction_a_cell was wrong, which results in incorrect weight files

@taehoonlee what was the error over here? I see you commit only renames the block. That can't be the source of the weight loading issue. 

I was able to get 3.8% error on the validation set as well, as well as  match he TF Hub scores with 1e-7 rtol for ImageNet Large.

I am having some difficulties with porting the NASNet Mobile weights, which are constantly producing the same erroneous predictions. I will look more closely at them shortly. 

For now, I am pushing the weights of the NASNet Large weights to my repository at https://github.com/titu1994/Keras-NASNet/releases/tag/v1.1

Since the files are so large, please give it 30+ mins to be uploaded.

File: keras/applications/nasnet.py [hunk 1]
```diff
@@ -210,24 +210,18 @@ def NASNet(input_shape=None,
     channel_dim = 1 if K.image_data_format() == 'channels_first' else -1
     filters = penultimate_filters // 24
 
-    if not skip_reduction:
```
Comment (titu1994) [inline keras/applications/nasnet.py]: I recommend not removing these if blocks. Without these checks, CIFAR models will be reduced from 32x32 to 4x4 prior to the first normal_a cell.
Instead : 
1) Rename `skip_reduction` to `skip_reduction_input_layer` to match TF
2) Set the `if not skip_reduction` to `if skip_reduction_input_layer` to allow only the Large model to use them. 
3) Set the NASNet Mobile version to **not** use `skip_reduction_input_layer`, according to the TF code.

**Comment by titu1994** [comment]
@taehoonlee Was able to fix the NASNet Mobile weights loading by adding another hyperparameter. I am posting the NASNet builder model below with comments pointing to the required changes. 

Would you replicate these changes to this PR?

```python
def NASNet(input_shape=None,
           penultimate_filters=4032,
           nb_blocks=6,
           stem_filters=96,
           initial_reduction=True,  <---- RENAME `skip_reduction` to `initial_reduction`
           skip_reduction_layer_input=True,   <------ New parameter which is analogous to the TF variable of same name
           use_auxiliary_branch=False,
           filters_multiplier=2,
           dropout=0.5,
           weight_decay=5e-5,
           include_top=True,
           weights=None,
           input_tensor=None,
           pooling=None,
           classes=1000,
           default_size=None):
    """
   Reducing the comments to only the necessary updates 

    # Arguments
 
        initial_reduction: Whether to add the reduction step at the beginning
            end of the network.
        skip_reduction_layer_input: Determines whether to skip the reduction layers
            when calculating the previous layer to connect to.
    """
    if K.backend() != 'tensorflow':
        raise RuntimeError('Only Tensorflow backend is currently supported, '
                           'as other backends do not support '
                           'separable convolution.')

    if weights not in {'imagenet', None}:
        raise ValueError('The `weights` argument should be either '
                         '`None` (random initialization) or `imagenet` '
                         '(pre-training on ImageNet).')

    if weights == 'imagenet' and include_top and classes != 1000:
        raise ValueError('If using `weights` as ImageNet with `include_top` '
                         'as true, `classes` should be 1000')

    if default_size is None:
        default_size = 331

    # Determine proper input shape and default size.
    input_shape = _obtain_input_shape(input_shape,
                                      default_size=default_size,
                                      min_size=32,
                                      data_format=K.image_data_format(),
                                      require_flatten=include_top or weights)

    if K.image_data_format() != 'channels_last':
        warnings.warn('The NASNet family of models is only available '
                      'for the input data format "channels_last" '
                      '(width, height, channels). '
                      'However your settings specify the default '
                      'data format "channels_first" (channels, width, height).'
                      ' You should set `image_data_format="channels_last"` '
                      'in your Keras config located at ~/.keras/keras.json. '
                      'The model being returned right now will expect inputs '
                      'to follow the "channels_last" data format.')
        K.set_image_data_format('channels_last')
        old_data_format = 'channels_first'
    else:
        old_data_format = None

    if input_tensor is None:
        img_input = Input(shape=input_shape)
    else:
        if not K.is_keras_tensor(input_tensor):
            img_input = Input(tensor=input_tensor, shape=input_shape)
        else:
            img_input = input_tensor

    assert penultimate_filters % 24 == 0, "`penultimate_filters` needs to be divisible " \
                                          "by 24."

    channel_dim = 1 if K.image_data_format() == 'channels_first' else -1
    filters = penultimate_filters // 24

    """ This if line can be replaced by your PR if CIFAR is not to be supported """

    if initial_reduction:
        x = Conv2D(stem_filters, (3, 3), strides=(2, 2), padding='valid', use_bias=False, name='stem_conv1',
                   kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay))(img_input)
    else:
        x = Conv2D(stem_filters, (3, 3), strides=(1, 1), padding='same', use_bias=False, name='stem_conv1',
                   kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay))(img_input)

    x = BatchNormalization(axis=channel_dim, momentum=_BN_DECAY, epsilon=_BN_EPSILON,
                           name='stem_bn1')(x)

    p = None
   
    """ This if line can be removed if CIFAR is not to be supported """

    if initial_reduction:  # imagenet / mobile mode
        x, p = _reduction_A(x, p, filters // (filters_multiplier ** 2), weight_decay, id='stem_1')
        x, p = _reduction_A(x, p, filters // filters_multiplier, weight_decay, id='stem_2')

    for i in range(nb_blocks):
        x, p = _normal_A(x, p, filters, weight_decay, id='%d' % (i))

    x, p0 = _reduction_A(x, p, filters * filters_multiplier, weight_decay, id='reduce_%d' % (nb_blocks))

    """ This now uses the new attribute to determine the input connection """

    p = p0 if not skip_reduction_layer_input else p

    for i in range(nb_blocks):
        x, p = _normal_A(x, p, filters * filters_multiplier, weight_decay, id='%d' % (nb_blocks + i + 1))

    auxiliary_x = None
    if not initial_reduction:  # imagenet / mobile mode
        if use_auxiliary_branch:
            auxiliary_x = _add_auxiliary_head(x, classes, weight_decay, pooling, include_top)

    x, p0 = _reduction_A(x, p, filters * filters_multiplier ** 2, weight_decay, id='reduce_%d' % (2 * nb_blocks))

    if initial_reduction:  # CIFAR mode
        if use_auxiliary_branch:
            auxiliary_x = _add_auxiliary_head(x, classes, weight_decay, pooling, include_top)

     """ This now uses the new attribute to determine the input connection """

    p = p0 if not skip_reduction_layer_input else p

    for i in range(nb_blocks):
        x, p = _normal_A(x, p, filters * filters_multiplier ** 2, weight_decay, id='%d' % (2 * nb_blocks + i + 1))

    x = Activation('relu')(x)

    if include_top:
        x = GlobalAveragePooling2D()(x)
        x = Dropout(dropout)(x)
        x = Dense(classes, activation='softmax', kernel_regularizer=l2(weight_decay), name='predictions')(x)
    else:
        if pooling == 'avg':
            x = GlobalAveragePooling2D()(x)
        elif pooling == 'max':
            x = GlobalMaxPooling2D()(x)

    # Ensure that the model takes into account
    # any potential predecessors of `input_tensor`.
    if input_tensor is not None:
        inputs = get_source_inputs(input_tensor)
    else:
        inputs = img_input

    # Create model.
    if use_auxiliary_branch:
        model = Model(inputs, [x, auxiliary_x], name='NASNet_with_auxiliary')
    else:
        model = Model(inputs, x, name='NASNet')

    # Code to load weight not shown as it is different from Keras paths
```

After this, the following changes have to be made to the `NASNetLarge` and `NASNetMobile` methods : 
```python
def NASNetLarge(input_shape=(331, 331, 3),
                dropout=0.5,
                weight_decay=5e-5,
                use_auxiliary_branch=False,
                include_top=True,
                weights='imagenet',
                input_tensor=None,
                pooling=None,
                classes=1000):

    return NASNet(input_shape,
                  penultimate_filters=4032,
                  nb_blocks=6,
                  stem_filters=96,
                  initial_reduction=True,  <---- This is set for ImageNet stem
                  skip_reduction_layer_input=True,   <---- This is set for Large only
                  use_auxiliary_branch=use_auxiliary_branch,
                  filters_multiplier=2,
                  dropout=dropout,
                  weight_decay=weight_decay,
                  include_top=include_top,
                  weights=weights,
                  input_tensor=input_tensor,
                  pooling=pooling,
                  classes=classes,
                  default_size=331)
```

```python
def NASNetMobile(input_shape=(224, 224, 3),
                 dropout=0.5,
                 weight_decay=4e-5,
                 use_auxiliary_branch=False,
                 include_top=True,
                 weights='imagenet',
                 input_tensor=None,
                 pooling=None,
                 classes=1000):

     return NASNet(input_shape,
                  penultimate_filters=1056,
                  nb_blocks=4,
                  stem_filters=32,
                  initial_reduction=True,  <--- This is set for ImageNet stem
                  skip_reduction_layer_input=False,  <--- This is Not Set for Mobile Models and CIFAR models
                  use_auxiliary_branch=use_auxiliary_branch,
                  filters_multiplier=2,
                  dropout=dropout,
                  weight_decay=weight_decay,
                  include_top=include_top,
                  weights=weights,
                  input_tensor=input_tensor,
                  pooling=pooling,
                  classes=classes,
                  default_size=224)

```

**Comment by titu1994** [comment]
@fchollet I've added the updated weights to the following tag - https://github.com/titu1994/Keras-NASNet/releases/edit/v1.2

This is because the Auxiliary branch is not being used in the Keras master version.

**Comment by taehoonlee** [comment]
Here are top-5 error rates with the first 1000 samples of ImageNet validation set:
 - `fchollet/.../v0.8/NASNet-large.h5` (md5: `a86ad97c98c056cf7bde18e84c2559dd`): 3.6%,
 - `fchollet/.../v0.8/NASNet-mobile.h5` (md5: `d9e0bd07170f9751f5bc96fe5594c3c9`): 32.8%,
 - `titu1994/.../v1.1/NASNet-large.h5` (md5: `c70d45c315836cb17946fe9db58247df`): 3.8%,
 - `titu1994/.../v1.0/NASNet-mobile.h5` (md5: `020fb642bf7360b370c678b08e0adf61`): 7.8%,
 - `titu1994/.../v1.2/NASNet-large.h5` (md5: `11577c9a518f0070763c2b964a382f17`): 3.5%,
 - `titu1994/.../v1.2/NASNet-mobile.h5` (md5: `020fb642bf7360b370c678b08e0adf61`): the same as `v1.0`.

To @titu1994:

1. As you said, renaming can't be the source of the weight loading issue. But I assume that incorrect names affected values during your weight translations, because the `NASNet-mobile.h5` with `d9e0bd...` has random values only for the weights `reduction_right1_`.

2. We need to focus only ImageNet models, because the `keras/applications` is assumed to be coupled with `decode_predictions` which returns ImageNet labels. Thus, no further changes such as adding `initial_reduction` are required. This PR with `v1.2` weights are enough to reproduce the results in the original paper.

To @fchollet:

The four lines `https://github.com/keras-team/keras/blob/master/keras/applications/nasnet.py#L63-L66` needs to be replaced with [the v1.2 files](https://github.com/titu1994/Keras-NASNet/releases/tag/v1.2). Additionally, it would be nice if md5 values are added.

**Comment by titu1994** [comment]
@taehoonlee Finished porting all the mobilenet weights, matched to 1e-5 rtol with tf-hub and 90% top 5 accuracy on first 1k validation samples.

@fchollet The weights are ready for download. Sorry for the wait. NASNet is a very difficult architecture to deal with.

**Comment by titu1994** [comment]
@taehoonlee Agreed with second points mentioned above. No need to add the extra parameter since Keras is focussing on ImageNet.

As to the weight naming, weight loading is done on a precise layer wise placement. An error of mismatched shape or kernel shape difference would be thrown in such a case if the error was impacted by the name.

In any case, more so than the weight porting mechanism, it was the interconnection of the `p = p0 if ... else p` part which caused the difficulty. It doesn't throw an error, shapes match, and the differences in scores are imperceptible on just 1 or 2 random images. I'll be careful in validation on the validation set of ImageNet next time.

**Comment by taehoonlee** [comment]
Thank you for the updates @titu1994!

**Comment by fchollet** [comment]
Thanks both of you for the update. Can we include the weight file addresses update in this PR? (as I understand, the files are already hosted on GitHub?)

**Comment by titu1994** [comment]
@fchollet Yes the weights are currently hosted at - https://github.com/titu1994/Keras-NASNet/releases/tag/v1.2

**Comment by taehoonlee** [comment]
Thanks for quick response!

**Comment by titu1994** [comment]
@taehoonlee thanks for the fix. It's been a problem for quite some time.

**Comment by ahundt** [comment]
Awesome, thanks everyone! I'll give this updated version a try. 

@titu1994 might the keras-contrib version be easy to update accordingly as well?

**Comment by titu1994** [comment]
Sorry for the delay. Yes, this patch can be applied to NASNet on Contrib with a few modifications to support CIFAR models. 

I'll try pushing the patch by tomorrow

**Comment by michaelmunie** [comment]
I believe there is still an error in the NASNetMobile weights. Loading the model with the top works but loading the model without the top fails.  Also, thank you so much for the work on this fix.  Before, trying to get good performance from NASNet was very difficult.

```
from keras_applications.nasnet import NASNetMobile, preprocess_input
base_model_top = NASNetMobile(weights='imagenet', include_top=True)
base_model_notop = NASNetMobile(weights='imagenet', include_top=False)
```

>  ValueError                                Traceback (most recent call last)
> <ipython-input-12-5c2992946dd3> in <module>()
> ----> 3 base_model_notop = NASNetMobile(weights='imagenet', include_top=False)
> 
> ~/tf/lib/python3.5/site-packages/keras_applications/nasnet.py in NASNetMobile(input_shape, include_top, weights, input_tensor, pooling, classes)
>     424                   pooling=pooling,
>     425                   classes=classes,
> --> 426                   default_size=224)
>     427 
>     428 
> 
> ~/tf/lib/python3.5/site-packages/keras_applications/nasnet.py in NASNet(input_shape, penultimate_filters, num_blocks, stem_block_filters, skip_reduction, filter_multiplier, include_top, weights, input_tensor, pooling, classes, default_size)
>     269                     cache_subdir='models',
>     270                     file_hash='1ed92395b5b598bdda52abe5c0dbfd63')
> --> 271             model.load_weights(weights_path)
>     272         elif default_size == 331:  # large version
>     273             if include_top:
> 
> ~/tf/lib/python3.5/site-packages/keras/engine/network.py in load_weights(self, filepath, by_name, skip_mismatch, reshape)
>    1178             else:
>    1179                 saving.load_weights_from_hdf5_group(
> -> 1180                     f, self.layers, reshape=reshape)
>    1181 
>    1182     def _updated_config(self):
> 
> ~/tf/lib/python3.5/site-packages/keras/engine/saving.py in load_weights_from_hdf5_group(f, layers, reshape)
>     899                          'containing ' + str(len(layer_names)) +
>     900                          ' layers into a model with ' +
> --> 901                          str(len(filtered_layers)) + ' layers.')
>     902 
>     903     # We batch weight value assignments in a single backend call
> 
> ValueError: You are trying to load a weight file containing 388 layers into a model with 382 layers.

**Pull request merged**